### ISSUE TITLE
EC pretraining prediction accuracy collapses after achieving 99+% accuracy. 

### ISSUE THREAD

**Issue by soumye** [cambridgeltl/ECNMT#2]
Hi, 
I am trying to run the ECPRETRAINING script with default hparams. The model is able to reach 99% accuracy in 4k epochs but after that the prediction accuracy collapses. I am unable to find any bug in the code that might be causing this. I have attached the training logs for the same

**Reaching 99%**
![image](https://user-images.githubusercontent.com/19409862/107156979-7cad5800-694f-11eb-8da0-3c5ea6c4f16a.png)

**Collapse to 0%**
![image](https://user-images.githubusercontent.com/19409862/107156990-8a62dd80-694f-11eb-8bff-aa807c87673c.png)

**Full log**
[log.log](https://github.com/cambridgeltl/ECNMT/files/5939780/log.log)

It would great if you could help resolve this issue. 
Thanks.

**Comment by yaoyiran** [comment]
Hi, here is a simple solution: (1) set the learning rate as 1e-4 rather than the default 1e-3; (2) set a minimum message length in models.py, e.g. replacing line 26 "if idx_ == 0:" with , say, "if idx_ <= 5:", "if idx_ <= 10:" or "if idx_ >= 0:".

PS. To reproduce the results in our paper, you can just use the default hyper-parameters and stop the training process when you get a model whose EC accuracy is satisfactory.

**Comment by soumye** [comment]
I was indeed able to get the EC accuracy to a satisfactory 99%,  which I guess is good enough for pre-training. But the instability of the model to maintain that and the eventual drift I think highlights the brittleness of the training. I was wondering if there was a fix to prevent the model from collapsing later on.

**Comment by yaoyiran** [comment]
I've already mentioned that such a "fix" really exists: you could just set the lr = 1e-4 and set a minimum message length. Here is my training log with the new hyper-parameters, where the EC prediction accuracy on our valid set is >= 99.95% and super stable from the beginning to epoch 251000 and I expect it to be stable afterwards. Please feel free to leave a comment if you still cannot get a stable training process. 


[log.train.txt](https://github.com/cambridgeltl/ECNMT/files/5960320/log.train.txt)
